### PR TITLE
gprinttyp: pretty printing for levels

### PR DESCRIPTION
**Pull request by Octachron**
This small PR a pretty-printer for type node level in the `Gprinttyp` debugging module. To make those as lightweight as possible and to circumvent the limited support of nested subscript and superscript in graphiz , the pretty printer uses unicode superscript to render levels as superscript the right corner of the node indices (eg `1531 ⁸`), for instance the problematic counter-example check in #13600 can be rendered as

0045-Unify2_expand--

### CODE REVIEW

File: typing/gprinttyp.ml [hunk 1]
```diff
@@ -139,16 +139,29 @@ module Index: sig
     | Main of int
     | Synthetic of int
     | Named_subnode of { id:int; synth:bool; name:string }
+  type level_and_scope = { level:int; scope: int }
+  type 'a desc = {
+    id: 'a;
+    color: Decoration.color option;
+    desc: Types.type_desc;
+    lvl:level_and_scope;
+  }
```
Comment (gasche) [inline typing/gprinttyp.ml:148]: I think that these new types could be declared outside the `Index` module, as they are morally orthogonal to the index logic. (But the current choice is okay.)

File: typing/gprinttyp.ml [hunk 2]
```diff
@@ -521,8 +540,48 @@ module Digraph = struct
...
+    | 1 -> "¹"
+    | 2 -> "²"
+    | 3 -> "³"
+    | 0 -> "⁰"
+    | 4 -> "⁴"
+    | 5 -> "⁵"
+    | 6 -> "⁶"
+    | 7 -> "⁷"
+    | 8 -> "⁸"
+    | _ -> "⁹"
```
Comment (gasche) [inline typing/gprinttyp.ml]: I would `| 9 -> ... | _ -> assert false` to detect bad callsites.

File: typing/gprinttyp.ml [hunk 3]
```diff
@@ -521,8 +540,48 @@ module Digraph = struct
...
+    if n < 10 then
+      superscript_digit ppf n
+    else begin
+      superscript ppf (n/10);
+      superscript_digit ppf (n mod 10)
+    end
+
+  let superscript_level ppf lvl =
+    (* avoid a dependency on Btype *)
+    if lvl = Ident.highest_scope then Format.pp_print_string ppf "ʷ"
```
Comment (gasche) [inline typing/gprinttyp.ml]: space then [combining infinity](https://www.fileformat.info/info/unicode/char/1ab2/index.htm) would also work.

File: typing/gprinttyp.ml [hunk 4]
```diff
@@ -521,8 +540,48 @@ module Digraph = struct
...
+    let d = match lvl with
+      | None -> labelf "<SUB>%a</SUB>" Pp.prettier_index id
+      | Some {Index.level; scope=0} ->
+          labelf "<SUB>%a %a</SUB>"
+            Pp.prettier_index id superscript_level level
+      | Some {Index.level; scope} ->
+          labelf "<SUB>%a %a⁺%a</SUB>"
+            Pp.prettier_index id
+            superscript_level level
+            superscript scope
```
Comment (gasche) [inline typing/gprinttyp.ml:584]: a `superscript_scope` to factor the do-nothing-on-0 logic would have been nice, but the code is fine as-is and I would keep it that way.

File: typing/gprinttyp.mli [hunk 5]
```diff
@@ -20,10 +20,10 @@
...
 A type node is printed as
 {[
-    .------------.
-    | <desc>  id |---->
-    |            |--->
-    .------------.
+    .-------------.
+    | <desc>  idˡᵛˡ |---->
+    |             |--->
+    .-------------.
```
Comment (gasche) [inline typing/gprinttyp.mli]: The ASCII-art alignment is broken with monospace fonts. Maybe just `id^lvl` in ASCII?

**Review by gasche — APPROVED** [review]
Nice, thanks! See minor review comments.

File: typing/gprinttyp.ml (hunk 3 above)
Comment (Octachron) [reply typing/gprinttyp.ml]: This is better indeed (at least while we still have only one generic level), thanks for the suggestion!

**Comment by gasche** [comment]
I'll let you decide when you are satisfied with your own post-review modifications and then (you) merge.

**Pull request merged**